### ISSUE TITLE
Bug: The automatic update client behaves unintentionally

### ISSUE THREAD

**Issue by meetrick** [cosmos/relayer#1111]
# The automatic update client behaves unintentionally.

## Problem description: 
- When starting the relayer, the auto-update client function works even without setting a time threshold. Under certain conditions, update_client is triggered for the src chain every block. This continues until the balance of the relayer key account is exhausted and it can no longer process packets.

<img width="1169" alt="image" src="https://user-images.githubusercontent.com/19239345/221098207-7398823c-7b5c-40e2-825f-ba6b447d04c5.png">

## Function flow: 
- In the code, the processLatestMessages() function first calls the assembleAndSendMessages() function. The assembleAndSendMessages() function calculates the duration time from the last consensus height of the dst chain based on the src chain and generates an update client packet for the dst chain if the duration time is greater than trust_period * 2/3. Then, the processLatestMessages() function swaps the src and dst and calls the assembleAndSendMessages() function again. This time, the duration time is calculated from the last consensus height of the src chain based on the dst chain, and an update client packet is generated for the src chain if the duration time is greater than trust_period * 2/3.

## Problem 1: 
- However, when obtaining the last consensus height of src and dst, the function does not consider all packets that are being transmitted through the channel. Instead, it retrieves the last consensus height that has been processed by the key account of the registered chain. Thus, the function seems to assume that there is only one relayer operator for each channel. In reality, packets are being exchanged in the channel, but if the relayer operator's key account cannot process packets during the trust period, the relayer will trigger an update client.

## Problem 2: 
- The problem we encountered is as follows. While packets are being processed on the src chain, another relayer operator is competing for packets on the dst chain and cannot process them for (trust period * 2/3) seconds. Based on the src chain, the client on the dst chain is expected to expire, so an update client is triggered on the src chain. However, based on the dst chain, the src chain is processing packets, so no update client is triggered on the dst chain. This leads to the relayer repeatedly triggering an update client on the src chain in the next block, assuming that the dst chain is still unable to process packets.

**Comment by meetrick** [comment]
# Solution Proposal
## Solution.
- The update_client should be able to observe and process the actual packets transmitted over a specific channel.
## W/A
- When update_client is executed, if packets processed by dst chain based on src chain are not present for (trust period * 2/3) duration, send an update_client message to dst chain. Modifying as follows may be appropriate.
```go
diff --git a/relayer/processor/path_processor_internal.go b/relayer/processor/path_processor_internal.go
index 6426454..235282c 100644
--- a/relayer/processor/path_processor_internal.go
+++ b/relayer/processor/path_processor_internal.go
@@ -844,7 +844,7 @@ func (pp *PathProcessor) assembleAndSendMessages(
                dst.trackProcessingPacketMessage(m)
        }

-       go pp.sendMessages(ctx, src, dst, &om, pp.memo)
+       go pp.sendMessages(ctx, dst, src, &om, pp.memo)

        return nil
 }
```

Please give your opinions.

**Comment by agouin** [comment]
Hi @meetrick! Thank you for your report. 

I believe there may be something mis-configured, as we only see this behavior if the relayer has not yet successfully queried the client trusting period. The query happens when the relayer first starts up, so there must either be an error with that query, or the query is responding with a very low trusting period.

This does indicate a bug in the relayer though, where failures to query the trusting period will lead to constant update clients being broadcasted. A fix for that is here #1113 , which will log out any errors when querying the client trusting periods, and will not attempt to send client updates if there has not yet been a successful trusting period query.

Please try out the relayer built from that PR. It should give more information in an error log on why the trusting period query is failing. Also, would you please share your config? 

> the function does not consider all packets that are being transmitted through the channel.

This is not true as the CosmosChainProcessor observes all IBC messages from the chains. Any client updates from other relayers will also be observed and taken into consideration.

> Modifying as follows may be appropriate.

This change would break the correlation that happens within the PathProcessor.

**Comment by meetrick** [comment]
@agouin I agree with you, Thanks
